File: ts/models/weapon.ts
import { Attribute } from './attribute'
import { PurchaseCost } from './purchase';
import { RangeType } from './range';
import { UpgradeCost } from './upgradeCost';

/**
 * Shows the minimum level required for the player to level up their weapon to a certain level.
 */
export interface WeaponLevelRequirement {
    // the weapon level
    level: number,
    // the minimum player level required to level up the weapon to this level
    minPlayerLevelRequired: number,
}

/**
 * Represents a weapon.
 */
export interface Weapon {
    // the weapon's name
    name: string,
    // the weapon's description
    description: string,
    // the rarity of the weapon
    rarity: WeaponRarity,
    // the weapon's range type
    rangeType: RangeType,
    // the cost to purchase this weapon (if it is a purchasable weapon)
    purchaseCost: PurchaseCost,
    // the weapon's attribute (i.e. characteristics)
    levelMechanics: WeaponLevelMechanics[],
}

/**
 * Lists all current weapon rarities.
 */
export enum WeaponRarity {
    Common = 'Common',
    Uncommon = 'Uncommon',
    Rare = 'Rare',
    Epic = 'Epic',
    Legendary = 'Legendary'
}

/**
 * Represents an instance of a weapon's attribute
 */
export interface WeaponAttribute {
    // the weapon's attribute
    attribute: Attribute,
    // the chance for the attribute to play out
    chance: number,
    // the weapon's attribute modifier
    modifier: WeaponAttributeModifier,
    // the radius of effect of the weapon's attribute
    radius: number,
}

/**
 * Represents a weapon's data level by level
 */
export interface WeaponLevelMechanics {
    // the weapon's level
    level: number,
    // the weapon's base damage at this level
    baseDamage: number,
    // the weapon's base attack range at this level
    baseAttackRange: number,
    // the weapon's base attack time at this level
    baseAttackTime: number,
    // the weapon's crit chance at this level
    critChance: number,
    // the weapon's attributes at this level
    attributes: WeaponAttribute[],
    // the weapon's upgrade cost at this level
    upgradeCost: UpgradeCost,
}

/**
 * Represents a weapon's attribute modifier.
 */
export interface WeaponAttributeModifier {
    type: WeaponAttributeModifierType,
    value: number,

}

/**
 * Lists all current weapon attribute modifier types.
 */
export enum WeaponAttributeModifierType {
    Percentage = 'Percentage',
    Number = 'Number',
}

File: rust/models/weapon.rs
use serde::{Serialize, Deserialize};

use super::{RangeType, Attribute};

/// Represents a weapon and its data.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct Weapon {
    /// the name of the weapon
    name: String,
    /// the weapon's description
    description: String,
    /// the weapon's level
    level: u8,
    /// the weapon's rarity
    rarity: WeaponRarity,
    /// the weapon's range type
    #[serde(rename = "rangeType")]
    range_type: RangeType,
    /// represents the weapon's attributes
    attributes: Vec<WeaponAttribute>,
    /// the weapon's base damage
    #[serde(rename = "baseDamage")]
    base_damage: f64,
    /// the weapon's base attack range
    #[serde(rename = "baseAttackRange")]
    base_attack_range: f64,
    /// the weapon's base attack speed
    #[serde(rename = "baseAttackSpeed")]
    base_attack_speed: f64,
    /// the weapon's crit chance
    #[serde(rename = "critChance")]
    crit_chance: f64,
}

/// Represents a weapon's attributes.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct WeaponAttribute {
    /// the weapon's attribute
    attribute: Attribute,
    /// the chance for the attribute to play out (in a ratio from 0 to 1)
    chance: f64,
    /// the attribute's modifier
    modifier: WeaponAttributeModifier,
    /// the attribute's radius of effect (in units)
    radius: f64,
}

/// A list of all available weapon rarities.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub enum WeaponRarity {
    #[serde(rename = "Common")]
    Common,
    #[serde(rename = "Uncommon")]
    Uncommon,
    #[serde(rename = "Rare")]
    Rare,
    #[serde(rename = "Epic")]
    Epic,
    #[serde(rename = "Legendary")]
    Legendary
}

/// Represents a weapon's attribute modifier.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct WeaponAttributeModifier {
    /// the type of modifier
    #[serde(rename = "modifierType")]
    modifier_type: WeaponAttributeModifierType,
    /// the value of the modifier
    value: f64,
}

/// A list of available weapon attribute modifier types.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub enum WeaponAttributeModifierType {
    /// the modifier is a percentage
    #[serde(rename = "Percentage")]
    Percentage,
    /// the modifier is either an integer, unsigned integer or a float
    #[serde(rename = "Number")]
    Number,
}
